core/lib/external_price_api/src/cmc_api.rs

```diff
@@ -0,0 +1,199 @@
+        }
+        // drop read lock
+
+        let response = self.get("/v1/cryptocurrency/map").send().await?;
+        let status = response.status();
+        if !status.is_success() {
+            return Err(anyhow::anyhow!(
+                "Http error while fetching token id. Status: {status}, token: {address}, msg: {}",
+                response.text().await.unwrap_or_default(),
+            ));
+        }
+
+        let parsed = response.json::<V1CryptocurrencyMapResponse>().await?;
+        for token_info in parsed.data {
+            if let Some(platform) = token_info.platform {
+                if platform.id == ALLOW_TOKENS_ONLY_ON_CMC_PLATFORM_ID
+                    && Address::from_str(&platform.token_address).is_ok_and(|a| a == address)
+                {
+                    self.cache_token_id_by_address
+                        .write()
+                        .await
+                        .insert(address, token_info.id);
+                    return Ok(token_info.id);
+                }
+            }
+        }
+
+        Err(anyhow::anyhow!("Token ID not found for address {address}"))
+    }
+
+    async fn get_token_price_by_address(&self, address: Address) -> anyhow::Result<f64> {
+        let id = self.get_token_id(address).await?;
+        self.get_token_price_by_id(id).await
+    }
+
+    async fn get_token_price_by_id(&self, id: i32) -> anyhow::Result<f64> {
+        let response = self
+            .get("/v2/cryptocurrency/quotes/latest")
+            .query(&[("id", id)])
+            .send()
+            .await?;
+
+        let status = response.status();
+        if !status.is_success() {
+            return Err(anyhow::anyhow!(
+                "Http error while fetching token price. Status: {status}, token: {id}, msg: {}",
+                response.text().await.unwrap_or("".to_string())
+            ));
+        }
+
+        response
+            .json::<V2CryptocurrencyQuotesLatestResponse>()
+            .await?
+            .data
+            .get(&id)
+            .and_then(|data| data.quote.get("USD"))
+            .map(|mq| mq.price)
+            .ok_or_else(|| anyhow::anyhow!("Price not found for token: {id}"))
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct V2CryptocurrencyQuotesLatestResponse {
+    data: HashMap<i32, CryptocurrencyQuoteObject>,
+}
+
+#[derive(Debug, Deserialize)]
+struct CryptocurrencyQuoteObject {
+    // #[serde(flatten)]
+    // cryptocurrency_object: CryptocurrencyObject,
+    quote: HashMap<String, MarketQuote>,
+}
+
+#[derive(Debug, Deserialize)]
+struct MarketQuote {
+    price: f64,
+}
+
+#[derive(Debug, Deserialize)]
+struct V1CryptocurrencyMapResponse {
+    data: Vec<CryptocurrencyObject>,
+}
+
+#[derive(Debug, Deserialize)]
+struct CryptocurrencyObject {
+    id: i32,
+    // name: String,
+    // symbol: String,
+    // slug: String,
+    // is_active: u8, // TODO: This field is available, should we at least emit a warning if the listing is not marked as active?
```

Comment:
Do we care to do "liveness" checks (i.e. could emit a warning if the API says that the token is inactive)?

Reply:
That's a good point - lets absolutely do it!